File: frontend/src/api/canvas.ts
import http from './http'

export interface CanvasProcess {
  process_id: string
  name: string
  channel?: string | null
  description?: string | null
  entrypoints?: string | null
}

export interface CanvasStep {
  step_id: string
  name?: string | null
  description?: string | null
  step_type?: string | null
}

export interface CanvasEdge {
  id?: number
  from_step_id: string
  to_step_id: string
  edge_type?: string | null
  condition?: string | null
  label?: string | null
  from_handle?: string | null
  to_handle?: string | null
}

export interface CanvasImplementation {
  impl_id: string
  name?: string | null
  type?: string | null
  system?: string | null
  description?: string | null
  code_ref?: string | null
}

export interface CanvasStepImplLink {
  id?: number
  step_id: string
  impl_id: string
  step_handle?: string | null
  impl_handle?: string | null
}

export interface CanvasDataResource {
  resource_id: string
  name?: string | null
  type?: string | null
  system?: string | null
  location?: string | null
  description?: string | null
}

export interface CanvasImplDataLink {
  id?: number
  impl_id: string
  resource_id: string
  impl_handle?: string | null
  resource_handle?: string | null
  access_type?: string | null
  access_pattern?: string | null
}

export interface CanvasImplLink {
  id?: number
  from_impl_id: string
  to_impl_id: string
  from_handle?: string | null
  to_handle?: string | null
  edge_type?: string | null
  condition?: string | null
  label?: string | null
}

export interface ProcessCanvas {
  process: CanvasProcess
  steps: CanvasStep[]
  edges: CanvasEdge[]
  implementations: CanvasImplementation[]
  step_impl_links: CanvasStepImplLink[]
  data_resources: CanvasDataResource[]
  impl_data_links: CanvasImplDataLink[]
  impl_links: CanvasImplLink[]
}

export async function getProcessCanvas(processId: string): Promise<ProcessCanvas> {
  const { data } = await http.get<ProcessCanvas>(`/canvas/${processId}`)
  return data
}

export async function saveProcessCanvas(
  processId: string,
  payload: ProcessCanvas,
): Promise<ProcessCanvas> {
  const { data } = await http.put<ProcessCanvas>(`/canvas/${processId}`, payload)
  return data
}

File: backend/app/schemas/canvas.py
from typing import List, Optional

from pydantic import BaseModel


class CanvasProcess(BaseModel):
    process_id: str
    name: Optional[str] = None
    channel: Optional[str] = None
    description: Optional[str] = None
    entrypoints: Optional[List[str]] = None


class CanvasStep(BaseModel):
    step_id: str
    name: Optional[str] = None
    description: Optional[str] = None
    step_type: Optional[str] = None


class CanvasEdge(BaseModel):
    id: Optional[int] = None
    from_step_id: str
    to_step_id: str
    edge_type: Optional[str] = None
    condition: Optional[str] = None
    label: Optional[str] = None
    from_handle: Optional[str] = None
    to_handle: Optional[str] = None


class CanvasImplementation(BaseModel):
    impl_id: str
    name: Optional[str] = None
    type: Optional[str] = None
    system: Optional[str] = None
    description: Optional[str] = None
    code_ref: Optional[str] = None


class CanvasStepImplLink(BaseModel):
    id: Optional[int] = None
    step_id: str
    impl_id: str
    step_handle: Optional[str] = None
    impl_handle: Optional[str] = None


class CanvasDataResource(BaseModel):
    resource_id: str
    name: Optional[str] = None
    type: Optional[str] = None
    system: Optional[str] = None
    location: Optional[str] = None
    description: Optional[str] = None


class CanvasImplDataLink(BaseModel):
    id: Optional[int] = None
    impl_id: str
    resource_id: str
    impl_handle: Optional[str] = None
    resource_handle: Optional[str] = None
    access_type: Optional[str] = None
    access_pattern: Optional[str] = None


class CanvasImplLink(BaseModel):
    id: Optional[int] = None
    from_impl_id: str
    to_impl_id: str
    from_handle: Optional[str] = None
    to_handle: Optional[str] = None
    edge_type: Optional[str] = None
    condition: Optional[str] = None
    label: Optional[str] = None


class SaveProcessCanvasRequest(BaseModel):
    process_id: str
    process: CanvasProcess
    steps: List[CanvasStep] = []
    edges: List[CanvasEdge] = []
    implementations: List[CanvasImplementation] = []
    step_impl_links: List[CanvasStepImplLink] = []
    data_resources: List[CanvasDataResource] = []
    impl_data_links: List[CanvasImplDataLink] = []
    impl_links: List[CanvasImplLink] = []
